src/tmpfiles/tmpfiles.c

```diff
@@ -181,33 +191,81 @@ STATIC_DESTRUCTOR_REGISTER(arg_root, freep);
-        { 'b', specifier_boot_id,         NULL },
-        { 'H', specifier_host_name,       NULL },
-        { 'l', specifier_short_host_name, NULL },
-        { 'v', specifier_kernel_release,  NULL },
-        { 'a', specifier_architecture,    NULL },
-        { 'o', specifier_os_id,           NULL },
-        { 'w', specifier_os_version_id,   NULL },
-        { 'B', specifier_os_build_id,     NULL },
-        { 'W', specifier_os_variant_id,   NULL },
-
-        { 'g', specifier_group_name,      NULL },
-        { 'G', specifier_group_id,        NULL },
-        { 'U', specifier_user_id,         NULL },
-        { 'u', specifier_user_name,       NULL },
-        { 'h', specifier_user_home,       NULL },
-
-        { 't', specifier_directory,       UINT_TO_PTR(DIRECTORY_RUNTIME) },
-        { 'S', specifier_directory,       UINT_TO_PTR(DIRECTORY_STATE) },
-        { 'C', specifier_directory,       UINT_TO_PTR(DIRECTORY_CACHE) },
-        { 'L', specifier_directory,       UINT_TO_PTR(DIRECTORY_LOGS) },
-        { 'T', specifier_tmp_dir,         NULL },
-        { 'V', specifier_var_tmp_dir,     NULL },
-        {}
+typedef struct SpecifierForward {
+        Specifier specifier;
+        SpectableFlag flag;
+} SpecifierForward;
+
+static int disabled_specifier(char specifier, const void *data, const void *userdata, char **ret) {
+        return -ENXIO;
+}
+
+static const SpecifierForward specifier_table_data[] = {
+        { { 'm', specifier_machine_id_safe }, SPECTABLE_PREFIXARG },
+        { { 'b', specifier_boot_id },         SPECTABLE_RUNTIME },
+        { { 'H', specifier_host_name },       SPECTABLE_RUNTIME },
+        { { 'l', specifier_short_host_name }, SPECTABLE_RUNTIME },
+        { { 'v', specifier_kernel_release },  SPECTABLE_RUNTIME },
+        { { 'a', specifier_architecture },    SPECTABLE_RUNTIME },
+        { { 'o', specifier_os_id },           SPECTABLE_PREFIXARG },
+        { { 'w', specifier_os_version_id },   SPECTABLE_PREFIXARG },
+        { { 'B', specifier_os_build_id },     SPECTABLE_PREFIXARG },
+        { { 'W', specifier_os_variant_id },   SPECTABLE_PREFIXARG },
+
+        { { 'g', specifier_group_name },      SPECTABLE_USERDB },
+        { { 'G', specifier_group_id },        SPECTABLE_USERDB },
+        { { 'U', specifier_user_id },         SPECTABLE_USERDB },
+        { { 'u', specifier_user_name },       SPECTABLE_USERDB },
+        { { 'h', specifier_user_home },       SPECTABLE_USERDB },
+
+        { { 't', specifier_directory, UINT_TO_PTR(DIRECTORY_RUNTIME) }, SPECTABLE_DIRECTORY },
+        { { 'S', specifier_directory, UINT_TO_PTR(DIRECTORY_STATE) },   SPECTABLE_DIRECTORY },
+        { { 'C', specifier_directory, UINT_TO_PTR(DIRECTORY_CACHE) },   SPECTABLE_DIRECTORY },
+        { { 'L', specifier_directory, UINT_TO_PTR(DIRECTORY_LOGS) },    SPECTABLE_DIRECTORY },
+        { { 'T', specifier_tmp_dir,     "/tmp" },     SPECTABLE_STRING },
+        { { 'V', specifier_var_tmp_dir, "/var/tmp" }, SPECTABLE_STRING },
```

Comment:
not sure i follow here. Why not just define two `Specifier[]` tables, one for the off and one for the on case?

Reply:
The idea is to allow en/disabling groups of specifiers, and overwriting of specifiers via ENV variables. I ran into limitations with setting the hostname, as there are 2 related specifiers.
I am all ears if you can propose a good scheme.

Reply:
i don't get why you need these groups, though?

Reply:
First, I have the values from the default table and need to replace parts of it, so I already use this info (see table). Of course right now I could just use 2 tables, but...
But take for example the os-release or hostname specifiers, I planned that those could have 3-4 different states:

-   enabled (pick from running system)
-   disabled (skip lines containing those specs)
-   from file in root prefix
-   optionally overridden explicitly

grouping the specs should allow stuff like changing long + short hostname.

The other solution I considered is running all logic at start and building a table with `specifier_string` entries. Would be more straight up.

Reply:
i really don#t see the usecase for the complex modes. even if there's need for them eventually I'd start out simple and just do "on" and "off"

Reply:
Well, maybe you see now with the new version. I cant just add the prefix to a static table, I just don't see the simple solution you seem to have in mind.